waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,311 @@
+    maxFrameSize: int = DefaultMaxFrameSize,
+    syncInterval: timer.Duration = DefaultSyncInterval,
+    relayJitter: Duration = DefaultGossipSubJitter,
+    pruneCB: Option[PruneCallBack] = none(PruneCallback),
+    transferCB: Option[TransferCallback] = none(TransferCallback),
+): Result[T, string] =
+  let storage = Storage.new().valueOr:
+    return err("negentropy storage creation failed")
+
+  let sync = WakuSync(
+    storage: storage,
+    peerManager: peerManager,
+    maxFrameSize: maxFrameSize,
+    syncInterval: syncInterval,
+    relayJitter: relayJitter,
+    pruneCallBack: pruneCB,
+    pruneOffset: syncInterval div 2,
+    transferCallBack: transferCB,
+  )
+
+  sync.initProtocolHandler()
+
+  info "WakuSync protocol initialized"
+
+  return ok(sync)
+
+proc periodicSync(self: WakuSync, callback: TransferCallback) {.async.} =
+  debug "periodic sync initialized", interval = $self.syncInterval
+
+  while true:
+    await sleepAsync(self.syncInterval)
+
+    debug "periodic sync started"
+
+    let (hashes, peer) = (await self.sync()).valueOr:
+      debug "sync failed", error
+      continue
+
+    if hashes.len > 0:
+      (await callback(hashes, peer.peerId)).isOkOr:
+        debug "transfer callback failed", error
+        continue
+
+    debug "periodic sync done", hashSynced = hashes.len
+
+proc periodicPrune(self: WakuSync, callback: PruneCallback) {.async.} =
+  debug "periodic prune initialized", interval = $self.syncInterval
+
+  await sleepAsync(self.syncInterval)
+
+  var pruneStop = getNowInNanosecondTime()
+
+  while true:
+    await sleepAsync(self.syncInterval)
+
+    debug "periodic prune started",
+      startTime = self.pruneStart - self.pruneOffset.nanos,
+      endTime = pruneStop,
+      storageSize = self.storage.len
+
+    var (elements, cursor) =
+      (newSeq[(WakuMessageHash, Timestamp)](0), none(WakuMessageHash))
+
+    var tries = 3
+    while true:
+      (elements, cursor) = (
+        await callback(self.pruneStart - self.pruneOffset.nanos, pruneStop, cursor)
+      ).valueOr:
+        debug "pruning callback failed", error
+        if tries > 0:
+          tries -= 1
+          await sleepAsync(30.seconds)
+          continue
+        else:
+          break
+
+      if elements.len == 0:
+        break
+
+      for (hash, timestamp) in elements:
+        self.storage.erase(timestamp, hash).isOkOr:
+          trace "storage erase failed", timestamp, hash, error
+          continue
+
+      if cursor.isNone():
+        break
+
+    self.pruneStart = pruneStop
+    pruneStop = getNowInNanosecondTime()
+
+    debug "periodic prune done", storageSize = self.storage.len
+
+proc start*(self: WakuSync) =
+  self.started = true
+  self.pruneStart = getNowInNanosecondTime()
+
+  if self.transferCallBack.isSome() and self.syncInterval > ZeroDuration:
+    self.periodicSyncFut = self.periodicSync(self.transferCallBack.get())
+
+  if self.pruneCallBack.isSome() and self.syncInterval > ZeroDuration:
```

Comment:
I think there should always be a pruneCallback. If not, mem leak

Reply:
I agree but testing with pruning is very annoying so I made it configurable.